### PR TITLE
feat: CEL validation for provider-specific labels/requirements

### PR DESCRIPTION
**Pull request by tallaxes**
**Description**

This adds additional CEL rules for validating provider-specific labels in NodePools and requirements in NodePools/NodeClaims. These are added on the provider side to core CRDs. The actual impactful changes are in the CRDs, the rest is infra for generating the patches (and maintaining them in a readable way). The infra part can be improved, this puts just enough to capture the results.

It also removes `topology.kubernetes.io/zone` from allowed labels and requirements, until the continuous drift it causes is fixed.

**How was this change tested?**

* `make presumbit`
* manual testing of CRs
* (current testing framework need to be updated to cover provider-specific CEL validation)

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [ ] No

**Release Note**

```release-note

```

### CODE REVIEW

**Review by Bryce-Soghigian — APPROVED** [review]
> (current testing framework need to be updated to cover provider-specific CEL validation)

The only feedback here would be to add tests enforcing these changes. Seems you thought of that ahead of time in the pr. 

Approving given we come back and add tests when the framework supports it

**Review by Bryce-Soghigian — COMMENTED** [review]
/test

File: hack/validation/requirements.sh [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        "karpenter.azure.com/sku-accelerator",
+        "karpenter.azure.com/sku-networking-accelerated",
+        "karpenter.azure.com/sku-storage-premium-capable",
+        "karpenter.azure.com/sku-storage-ephemeralos-maxsize",
+        "karpenter.azure.com/sku-encryptionathost-capable",
+        "karpenter.azure.com/sku-gpu-name",
+        "karpenter.azure.com/sku-gpu-manufacturer",
+        "karpenter.azure.com/sku-gpu-count"
+    ]
+    || !self.find("^([^/]+)").endsWith("karpenter.azure.com")
```
Comment (Bryce-Soghigian) [inline hack/validation/requirements.sh:22]: > !self.find("^([^/]+)"

Can we leave a comment explaining this regex a bit? It is a bit magical to me. Rule applies for any well known labels and any labels that do not end with "karpenter.azure.com"? 


The () creates a grouping for extracting a string. then the slash means we want to extract a substring using that `/` as a back reference?

File: hack/validation/labels.sh [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+        "karpenter.azure.com/sku-networking-accelerated",
+        "karpenter.azure.com/sku-storage-premium-capable",
+        "karpenter.azure.com/sku-storage-ephemeralos-maxsize",
+        "karpenter.azure.com/sku-encryptionathost-capable",
+        "karpenter.azure.com/sku-gpu-name",
+        "karpenter.azure.com/sku-gpu-manufacturer",
+        "karpenter.azure.com/sku-gpu-count"
+    ]
+    || !x.find("^([^/]+)").endsWith("karpenter.azure.com")
+)
```
Comment (Bryce-Soghigian) [inline hack/validation/labels.sh:23]: rule is the same for both of these right? can they just be merged into one script?

File: hack/validation/requirements.sh (hunk 1 above)
Comment (tallaxes) [reply hack/validation/requirements.sh:22]: No, it just means "all non-slash characters after start" = "everything from start to the first /". I can add a comment ...

File: hack/validation/labels.sh (hunk 2 above)
Comment (tallaxes) [reply hack/validation/labels.sh:23]: It is close, but not quite the same, one operates on collection (of labels), another operates on a single thing (key). Yes, there are opportunities for sharing, that's what I mean by improving infra, but readability and result are more important for now.

**Review by Bryce-Soghigian — APPROVED** [review]

**Pull request merged**